### PR TITLE
refactor(select-utils): Migrate to typescript

### PR DESCRIPTION
**Pull request by Rhotimee**
#### Summary
Migrate select utils to typescript

### CODE REVIEW

File: packages/components/inputs/select-utils/src/clear-indicator/clear-indicator.tsx [hunk 1]
```diff
@@ -1,11 +1,19 @@
-import PropTypes from 'prop-types';
+import { CSSProperties, LegacyRef } from 'react';
```
Comment (emmenko) [inline packages/components/inputs/select-utils/src/clear-indicator/clear-indicator.tsx:1]: When importing types, prefer to use `import type`

```suggestion
import type { CSSProperties, LegacyRef } from 'react';
```

Comment (emmenko) [inline packages/components/inputs/select-utils/src/clear-indicator/clear-indicator.tsx:1]: When importing types, prefer to use `import type`

```suggestion
import type { CSSProperties, LegacyRef } from 'react';
```

File: packages/components/inputs/select-utils/src/wrapper-with-icon/wrapper-with-icon.tsx [hunk 2]
```diff
@@ -1,16 +1,29 @@
...
 import { customProperties } from '@commercetools-uikit/design-system';
-import { components as defaultComponents } from 'react-select';
+import {
+  components as defaultComponents,
+  SingleValueProps,
+  PlaceholderProps,
+} from 'react-select';
 
-const getDefaultComponent = (type) => {
+const getDefaultComponent = (type: string) => {
```
Comment (emmenko) [inline packages/components/inputs/select-utils/src/wrapper-with-icon/wrapper-with-icon.tsx:11]:
```suggestion
const getDefaultComponent = (type: TWrapperWithIconProps['type']) => {
```

File: .changeset/tough-eagles-visit.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@commercetools-uikit/select-utils': patch
+---
+
+Migrate to typescript
```
Comment (emmenko) [inline .changeset/tough-eagles-visit.md:5]: Let's also mention that `react-select` got updated to v5

File: packages/components/inputs/select-utils/package.json [hunk 4]
```diff
@@ -31,7 +31,7 @@
     "@emotion/styled": "^11.3.0",
     "lodash": "4.17.21",
     "prop-types": "15.7.2",
-    "react-select": "4.3.1"
+    "react-select": "^5.2.1"
```
Comment (emmenko) [inline packages/components/inputs/select-utils/package.json:34]: We need to update all other packages to v5 as well.

PS: we can maybe extract the upgrade to a separate PR.

**Comment by Rhotimee** [comment]
@emmenko, I am currently working on upgrading react-select in this [PR](https://github.com/commercetools/ui-kit/pull/1997).  That will automatically fix the issues.

File: .changeset/tough-eagles-visit.md (hunk 3 above)
Comment (Rhotimee) [reply .changeset/tough-eagles-visit.md:5]: This is already done in the react-select PR

File: packages/components/inputs/select-utils/src/create-select-styles.ts [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+  borderColor?: string;
+  boxShadow?: string;
+  pointerEvents?: string;
+};
+
+type TState = {
+  isFocused?: boolean;
+  isDisabled?: boolean;
+  isSelected?: boolean;
+};
```
Comment (emmenko) [inline packages/components/inputs/select-utils/src/create-select-styles.ts:44]: Is there a way to import those from react-select? For example https://github.com/JedWatson/react-select/blob/4e06d22e47bbd82a4f9f67452c7f16787c4ce529/packages/react-select/src/styles.ts#L52-L79

**Review by emmenko — COMMENTED** [review]
Overall looks good. Just have a question for the css styles.

File: packages/components/inputs/select-utils/src/create-select-styles.ts (hunk 5 above)
Comment (Rhotimee) [reply packages/components/inputs/select-utils/src/create-select-styles.ts:44]: For TBase, I tried using `CSSObjectWithLabel` but it caused the build to fail and I got this error.
`The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.` 

For TState, We can use `ClassNamesState` but the implementation is not explicit. So, I'm not sure it is a good idea to use it.

File: packages/components/inputs/select-utils/src/clear-indicator/clear-indicator.tsx [hunk 6]
```diff
@@ -1,11 +1,21 @@
...
+type TInnerProps = {
+  ref: LegacyRef<HTMLButtonElement>;
+} & JSX.IntrinsicElements['button'];
+
+type TClearIndicator = {
+  getStyles: (s: string, props: unknown) => CSSProperties;
+  innerProps: TInnerProps;
+} & ClearIndicatorProps;
+
+const ClearIndicator = (props: TClearIndicator) => {
```
Comment (emmenko) [inline packages/components/inputs/select-utils/src/clear-indicator/clear-indicator.tsx:18]: According to this example, we should be able to use `ClearIndicatorProps` directly. Can you double check?

https://github.com/JedWatson/react-select/blob/4e06d22e47bbd82a4f9f67452c7f16787c4ce529/docs/examples/CustomClearIndicator.tsx#L8-L13

**Review by emmenko — APPROVED** [review]
One small comment, otherwise I think this is good to go for now. Thanks!

File: packages/components/inputs/select-utils/src/clear-indicator/clear-indicator.tsx (hunk 6 above)
Comment (Rhotimee) [reply packages/components/inputs/select-utils/src/clear-indicator/clear-indicator.tsx]: Yes, we can use it directly. However, we need to update the ref type to use HTMLButtonElement since we are using a button instead of a div.
The `getStyles`  type def has been cleaned. 
Thanks.
